components/script/dom/readablebytestreamcontroller.rs

```diff
@@ -2,29 +2,159 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+use std::cell::{Cell, RefCell};
+use std::collections::VecDeque;
+
 use dom_struct::dom_struct;
+use js::jsapi::{Heap, JS_ClearPendingException};
+use js::jsval::{ObjectValue, UndefinedValue};
+use js::rust::wrappers::JS_GetPendingException;
 use js::rust::HandleValue as SafeHandleValue;
-use js::typedarray::ArrayBufferViewU8;
+use js::typedarray::{ArrayBufferU8, ArrayBufferViewU8};
 
 use super::bindings::buffer_source::HeapBufferSource;
+use super::bindings::cell::DomRefCell;
 use super::bindings::codegen::Bindings::ReadableStreamBYOBReaderBinding::ReadableStreamBYOBReaderReadOptions;
+use super::bindings::reflector::reflect_dom_object;
+use super::readablestream::ReaderType;
 use super::readablestreambyobreader::ReadIntoRequest;
-use super::types::ReadableStreamBYOBRequest;
+use super::readablestreamdefaultreader::ReadRequest;
+use crate::dom::bindings::buffer_source::{create_uint8_array_with_buffer, BufferSource};
 use crate::dom::bindings::codegen::Bindings::ReadableByteStreamControllerBinding::ReadableByteStreamControllerMethods;
 use crate::dom::bindings::import::module::{Error, Fallible};
-use crate::dom::bindings::reflector::Reflector;
+use crate::dom::bindings::reflector::{DomGlobal, Reflector};
 use crate::dom::bindings::root::{DomRoot, MutNullableDom};
+use crate::dom::bindings::trace::RootedTraceableBox;
+use crate::dom::globalscope::GlobalScope;
 use crate::dom::readablestream::ReadableStream;
+use crate::dom::readablestreambyobrequest::ReadableStreamBYOBRequest;
 use crate::script_runtime::{CanGc, JSContext as SafeJSContext};
 
+/// <https://streams.spec.whatwg.org/#readable-byte-stream-queue-entry>
+#[derive(JSTraceable, MallocSizeOf)]
+pub(crate) struct QueueEntry {
+    /// <https://streams.spec.whatwg.org/#readable-byte-stream-queue-entry-buffer>
+    #[ignore_malloc_size_of = "HeapBufferSource"]
+    buffer: HeapBufferSource<ArrayBufferU8>,
+    /// <https://streams.spec.whatwg.org/#readable-byte-stream-queue-entry-byte-offset>
+    byte_offset: RefCell<usize>,
```

Comment:
Is it possible to make these not part of a refcell, since the queue itself is in a refcell?

Reply:
In addition, cheap `Copy` types (like `usize`) should use a `Cell` instead., which is faster.